### ISSUE TITLE
Use Vec() instead of complex number for non-complex SIEs.

### ISSUE THREAD

**Issue by dlfivefifty** [ApproxFun/SingularIntegralEquations.jl#80]
ApproxFun now supports (for the most part) 1D domains embedded in R^2.  This should be used here, partly to draw a distinction between `dz` and `ds`, which are still confusing.

In this case, whether it's `dz` or `ds` would be inferred from the domain:

```julia
    d_2d=Interval((0,1),(1,1))   
    d_C=Interval(im,1+im)
     SingularIntegral(d_2d,1) # would be in R^2, so use ds as dz makes no sense
     SingularIntegral(d_C,1) # use dz
```

This requires a bit of work in ApproxFun before it can be used.  For example, I'd want the following to work:

```julia
s=Fun(d_2d)  # this is a Fun whose values are Vecs, and plays the role of z
s_0=Vec(1,2)   # plays the role of z_0=1+2im

log(norm(s-s_0))   # plays the role of log(abs(z-z_0))
```

**Comment by MikaelSlevinsky** [comment]
I think this will be an improvement. Currently, we are using complex numbers to represent (i) the real & imaginary parts of a `GreensFun`, (ii) the geometry, and (iii) use in vector-valued line integrals for the Neumann problem.

Complex numbers in Julia are Julian types, so I don't expect this will significantly alter performance. Instead, it will bring clarity to the situations above. It will also generalize to 3D.

We should be able to distinguish between a line integral over a scalar field (∫ f ds) and a vector field (∫ F ⋅ dr).

**Comment by MikaelSlevinsky** [comment]
Perhaps the last part is a question...

**Comment by dlfivefifty** [comment]
`Complex{Float64}` is not (just) a Julia type, it's stored as bits.  This is why BLAS is fast for Complex numbers.   But so is `Vec{2,Float64}`, with the exact same number of bits.  Therefore anywhere complex numbers are faster, we can convert between the two:

```julia
c=rand(Complex128,20)
v=unsafe_wrap(Array,reinterpret(Ptr{Vec{2,Float64}},pointer(c)),20)
v[1]=Vec(1,2)
c[1]  # returns 1.0+2.0im
```

**Comment by dlfivefifty** [comment]
(Note that `reinterpret(UInt128,1.0+2.0im)` does not work since `Complex128` is not a `bitstype`, even though `isbits(1.0+2.0im)` is true.  )

**Comment by MikaelSlevinsky** [comment]
Gotcha.

**Comment by daanhb** [comment]
Slightly off-topic, sorry, but does Vec refer to FixedSizeArrays? I've recently switched to using StaticArrays (following [this discussion](https://github.com/SimonDanisch/FixedSizeArrays.jl/issues/159)) for small vectors in 2D-3D. It is often convenient in code that the vectors in StaticArrays inherit from AbstractArray, whereas those in FixedSizeArrays don't. The constructors are also more friendly. (On the other hand, tracing dependencies in ApproxFun it seems FixedSizeArrays are imported via Plots, so switching may not be an option.)

**Comment by dlfivefifty** [comment]
Yep, I was referring to FixedSizeArrays.  It looks from that link that even @SimonDanisch agrees that StaticArrays is the way forward, and after spending a day adding special cases for Vec to ApproxFun, I also agree.

Though it's 0.5 only so let's postpone this change til 0.4 support is dropped (which should wait til the type inference big is fixed)

**Comment by daanhb** [comment]
I see, I did not realize the 0.5 restriction of StaticArrays.
